docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md

```diff
@@ -4,39 +4,173 @@ description: How to write and deploy a dApp to Oasis Emerald
+The Emerald ParaTime is deployed similarly: the [Emerald Mainnet] is deployed on the Oasis Mainnet
+network while the [Emerald Testnet] on the Oasis Testnet network. The Emerald state on the Mainnet
+is stable. Testnet, apart from running the unstable version of the code and being prone to bugs, can
+have the state deliberately wiped either on the Emerald ParaTime layer or on the Oasis Testnet
+network level, so you should **never deploy a production service on the Testnet**!
 
-## Oasis consensus network vs Emerald
+:::tip
+
+For testing purposes, visit our [Testnet faucet] to obtain some TEST which you
+can then use on the Emerald Testnet to pay for gas fees. The faucet supports
+sending TEST both to your Consensus layer address or to your address inside the
+ParaTime.
+
+:::
 
-Emerald is running inside the Oasis ParaTime. While the Emerald's blockchain is separated from the Oasis network consensus layer, they share two important keypoints. Both have synchronized blocks with a common block time (around 6 seconds) which means that Emerald transactions will require at least this amount of time to be confirmed. Secondly, they share the same native token - ROSE on Mainnet and TEST on Testnet. This means that the gas fees will be paid in ROSE/TEST.
+[network-parameters]: ../../oasis-network/network-parameters.md
+[Testnet]: ../../foundation/testnet/README.md
+[Emerald Mainnet]: ./README.mdx#mainnet
+[Emerald Testnet]: ./README.mdx#testnet
 
-On the other hand, the Oasis addresses are bech32-encoded (e.g. `oasis1qpupfu7e2n6pkezeaw0yhj8mcem8anj64ytrayne`) while Emerald uses the Ethereum-compatible hex-encoded addresses (e.g. `0x90adE3B7065fa715c7a150313877dF1d33e777D5`). This is not just a stylistic change - the underlying algorithm for signing the transactions is [ed25519](https://en.wikipedia.org/wiki/EdDSA#Ed25519) on Oasis and [ECDSA](https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm) on Emerald (the same as on the Ethereum network). For this reason we cannot use Oasis accounts directly in Emerald, but we need to use a special mechanism described in the [How to transfer ROSE into Emerald ParaTime](../../manage-tokens/how-to-transfer-rose-into-emerald-paratime.mdx) chapter. To deploy your smart contract on the Emerald Mainnet, you will have to perform this step.
+## Running Private Oasis Network Locally
+
+For convenient development and testing of your dApps the Oasis team prepared a
+docker image which brings you a complete Oasis stack to your desktop. This
```

Comment:
`prepared a docker image` -> `prepared the [oasisprotocol/emerald-dev](link to Docker Hub) Docker image`

Reply:
Also move the link to the bottom so you can reference it multiple times in the text.